### PR TITLE
Added Monoid and Applicative Instances

### PR DESCRIPTION
**Pull request by octopuscabbage**

### CODE REVIEW

**Comment by octopuscabbage** [comment]
This addresses part of #17

File: Data/Matrix.hs [hunk 1]
```diff
@@ -152,6 +153,47 @@ instance Functor Matrix where
...
+instance Monoid a => Monoid (Matrix a) where
+  mempty = fromList 1 1 [mempty] 
+  mappend m m' = matrix (max (nrows m) (nrows m')) (max (ncols m) (ncols m')) $ uncurry zipTogether
+    where zipTogether row column
+            | (isJust melem && isJust m'elem) = (fromJust melem) <> (fromJust m'elem)
+            | (isJust melem) = (fromJust melem)
+            | (isJust m'elem) = fromJust m'elem
+            | otherwise = mempty
+                          where melem = safeGet row column m
+                                m'elem = safeGet row column m'
```
Comment (Daniel-Diaz) [inline Data/Matrix.hs]: Why not using the monoid instance for ``Maybe``?

```
fromMaybe mempty $ safeGet row column m <> safeGet row column m'
```

File: Data/Matrix.hs [hunk 2]
```diff
@@ -152,6 +153,47 @@ instance Functor Matrix where
...
+
+-------------------------------------------------------
+-------------------------------------------------------
+---- APPLICATIVE INSTANCE
+---- Works like tensor product but applies a function 
+
+instance Applicative Matrix where
+  pure x = fromList 1 1 [x] 
+  m <*> m' = flatten $ ((\f -> f <$> m') <$> m)
+
```
Comment (Daniel-Diaz) [inline Data/Matrix.hs]: Does this instance obey the ``Applicative`` class laws? (the same could be asked about the monoid instance above)

File: Data/Matrix.hs [hunk 3]
```diff
@@ -152,6 +153,47 @@ instance Functor Matrix where
...
+-------------------------------------------------------
+-------------------------------------------------------
+
+
+
+-- | Flatten a matrix of matrices. All sub matrices must have same dimensions
+--   This criteria is not checked. 
+flatten:: (Matrix (Matrix a)) -> Matrix a
+flatten m = foldl1 (<->) $ map (foldl1 (<|>)) $ map (\i -> getRow i m) [1..(nrows m)]
```
Comment (Daniel-Diaz) [inline Data/Matrix.hs]: This is a great addition, thanks.

One thing though: ``map f . map g = map (f . g)``. This could be used to use ``map`` only once.

**Comment by Daniel-Diaz** [comment]
Could you make the changes in tests a separate PR? I would like to review that independently.

Thanks for contributing!

File: Data/Matrix.hs (hunk 2 above)
Comment (octopuscabbage) [inline Data/Matrix.hs]: As far as I can tell yes. I worked some examples on paper but I couldn't get quickcheck to quite work for the applicative instances.

**Comment by octopuscabbage** [comment]
Do you mean the structural changes or changing the format of them? Because the format is already in this PR #21 

I should've added that this one was downstream to that one.

File: Data/Matrix.hs (hunk 1 above)
Comment (octopuscabbage) [inline Data/Matrix.hs]: Good idea, fixed that

File: Data/Matrix.hs (hunk 3 above)
Comment (octopuscabbage) [inline Data/Matrix.hs]: Fixed

**Comment by Daniel-Diaz** [comment]
I know you opened that PR, but the changes are in this PR as well. Check the "Files Changed" tab.

**Comment by Daniel-Diaz** [comment]
I checked the ``Applicative`` laws myself. They all seem to hold. Beautiful.

**Comment by octopuscabbage** [comment]
I think I get what you want. You want me to take the tests out of this and make them their own PR which would be downstream from the other test PR? 

Perhaps someone smarter than me can attempt to get the quickcheck tests working for applicative, as someone smarter than me told me what the applicative instance should be.

There was some talk of a monad instance which I guess would be similar to the applicative instance except flattened twice but that one is a lot harder to test to make sure the monad laws work. I guess it would work similarly to how the list monad works.

**Comment by octopuscabbage** [comment]
Alright now this can be merged independently of #21 and #28

**Comment by Daniel-Diaz** [comment]
Thanks a lot!

**Comment by octopuscabbage** [comment]
np, always happy to contrib

**Comment by Daniel-Diaz** [comment]
**About the monad instance**

To write a monad instance we will have a function the same type as flatten:

```haskell
join :: Matrix (Matrix a) -> Matrix a
```

With the submatrices being allowed to have any size. I don't see a way to implement that in any reasonable way.

**Comment by octopuscabbage** [comment]
This is the error I ran into. I don't think a monad is possible.

**Pull request merged**